game-core/src/main/java/games/strategy/triplea/attachments/TechAbilityAttachment.java

```diff
@@ -102,25 +109,72 @@ public TechAbilityAttachment(final String name, final Attachable attachable, fin
     super(name, attachable, gameData);
   }
 
-  // setters and getters
+  private UnitType getUnitType(final String value) throws GameParseException {
+    return Optional.ofNullable(getData().getUnitTypeList().getUnitType(value))
+        .orElseThrow(() -> new GameParseException("No unit called:" + value + thisErrorMsg()));
+  }
+
+  private String[] splitAndValidate(final String name, final String value) throws GameParseException {
+    final String[] stringArray = value.split(":");
+    if (stringArray.length <= 0 || stringArray.length > 2) {
+      throw new GameParseException(
+          String.format("%s cannot be empty or have more than two fields %s", name, thisErrorMsg()));
+    }
+    return stringArray;
+  }
+
+  private void applyCheckedValue(
+      final String name,
+      final String value,
+      final BiConsumer<UnitType, Integer> putter) throws GameParseException {
+    final String[] s = splitAndValidate(name, value);
+    putter.accept(getUnitType(s[1]), getInt(s[0]));
+  }
+
+  private static int sumIntegerMap(final Function<TechAbilityAttachment, IntegerMap<UnitType>> mapper,
+      final UnitType ut,
+      final PlayerID player,
+      final GameData data) {
+    return TechTracker.getCurrentTechAdvances(player, data).stream()
+        .map(TechAbilityAttachment::get)
+        .filter(Objects::nonNull)
+        .map(mapper)
+        .mapToInt(m -> m.getInt(ut))
+        .sum();
+  }
+
+  private static int sumNumbers(
+      final ToIntFunction<TechAbilityAttachment> mapper,
+      final PlayerID player,
+      final GameData data) {
+    return TechTracker.getCurrentTechAdvances(player, data).stream()
+        .map(TechAbilityAttachment::get)
+        .filter(Objects::nonNull)
+        .mapToInt(mapper)
+        .filter(i -> i > 0)
+        .sum();
+  }
+
+  private int getIntInRange(final String name, final String value, final int max, final boolean allowUndefined)
+      throws GameParseException {
+    final int intValue = getInt(value);
+    if (intValue < (allowUndefined ? -1 : 0) || intValue > max) {
+      throw new GameParseException(String.format(
+          "%s must be%s between 0 and %s%s",
```

Comment:
Should there be spaces between the format specifiers?  Or, if not for the second one, should the argument below have a leading space before `-1`?

Reply:
Right, seems like I accidentally removed it